### PR TITLE
SVS-109: Solution Items folder created during binding should be named…

### PR DESCRIPTION
**Pull request by Evangelink**
… SonarQube

### CODE REVIEW

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [ProjectSystemHelper.cs#L60](https://github.com/SonarSource-VisualStudio/sonarlint-visualstudio/blob/284bb793983c0d934ff576593348ba7b586fc801/src/Integration/LocalServices/ProjectSystemHelper.cs#L60): Consider using "Equals" if value comparison was intended. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1698)
1. MAJOR [ProjectSystemHelper.cs#L82](https://github.com/SonarSource-VisualStudio/sonarlint-visualstudio/blob/284bb793983c0d934ff576593348ba7b586fc801/src/Integration/LocalServices/ProjectSystemHelper.cs#L82): Remove this useless assignment to local variable "project". [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1854)
1. MAJOR [ProjectSystemHelper.cs#L266](https://github.com/SonarSource-VisualStudio/sonarlint-visualstudio/blob/284bb793983c0d934ff576593348ba7b586fc801/src/Integration/LocalServices/ProjectSystemHelper.cs#L266): Remove this useless assignment to local variable "hr". [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1854)
1. MAJOR [ProjectSystemHelper.cs#L295](https://github.com/SonarSource-VisualStudio/sonarlint-visualstudio/blob/284bb793983c0d934ff576593348ba7b586fc801/src/Integration/LocalServices/ProjectSystemHelper.cs#L295): Remove this useless assignment to local variable "hr". [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1854)
1. MAJOR [ProjectSystemHelper.cs#L320](https://github.com/SonarSource-VisualStudio/sonarlint-visualstudio/blob/284bb793983c0d934ff576593348ba7b586fc801/src/Integration/LocalServices/ProjectSystemHelper.cs#L320): Remove this useless assignment to local variable "hr". [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1854)

File: src/Integration.UnitTests/LocalServices/ProjectSystemHelperTests.cs [hunk 1]
```diff
@@ -137,21 +137,76 @@ public void ProjectSystemHelper_IsFileInProject_DifferentCase()
...
+        [TestMethod]
+        public void GetSolutionFolderProject_WhenCalledMultipleTimes_ReturnsSameProject()
+        {
+            // Setup
+            var solutionFolderName = "SomeFolderName";
+            DTEMock dte = new DTEMock();
+            this.serviceProvider.RegisterService(typeof(DTE), dte);
+            dte.Solution = this.solutionMock;
+            Project project1 = this.testSubject.GetSolutionFolderProject(solutionFolderName, true);
+
```
Comment (tvsonar) [inline src/Integration.UnitTests/LocalServices/ProjectSystemHelperTests.cs]: Extra empty line should be removed.

File: src/Integration/Binding/SolutionBindingOperation.cs [hunk 2]
```diff
@@ -197,6 +197,7 @@ public bool CommitSolutionBinding()
                 {
                     Debug.Assert(this.sourceControlledFileSystem.FileExist(info.NewRuleSetFilePath), "File not written " + info.NewRuleSetFilePath);
                     this.AddFileToSolutionItems(info.NewRuleSetFilePath);
+                    this.RemoveFileFromSolutionItems(info.NewRuleSetFilePath);
```
Comment (tvsonar) [inline src/Integration/Binding/SolutionBindingOperation.cs]: This method name is strange after reading the preceding line. Let's change it to something like `RemoveFileFromOldSolutionItems`

File: src/Integration/Binding/SolutionBindingOperation.cs [hunk 3]
```diff
@@ -256,6 +257,20 @@ private void AddFileToSolutionItems(string fullFilePath)
             }
         }
 
+        private void RemoveFileFromSolutionItems(string fullFilePath)
+        {
+            Debug.Assert(Path.IsPathRooted(fullFilePath) && this.sourceControlledFileSystem.FileExist(fullFilePath), "Expecting a rooted path to existing file");
+
+            Project solutionItemsProject = this.projectSystem.GetSolutionItemsProject(false);
```
Comment (tvsonar) [inline src/Integration/Binding/SolutionBindingOperation.cs]: Looking at this line it's not clear what `false` is. Adding `createOnNull: false` would help readability a lot.

File: src/Integration/Binding/SolutionBindingOperation.cs [hunk 4]
```diff
@@ -256,6 +257,20 @@ private void AddFileToSolutionItems(string fullFilePath)
...
+            if (solutionItemsProject != null)
+            {
+                // Remove file from project and if project is empty, remove project from solution
+                var fileName = Path.GetFileName(fullFilePath);
+                this.projectSystem.RemoveFileFromProject(solutionItemsProject, fileName);
+
```
Comment (tvsonar) [inline src/Integration/Binding/SolutionBindingOperation.cs]: extra empty line should be removed.

File: src/Integration/LocalServices/ProjectSystemHelper.cs [hunk 5]
```diff
@@ -186,6 +185,25 @@ public void AddFileToProject(Project project, string fullFilePath, string itemTy
             }
         }
 
+        public void RemoveFileFromProject(Project project, string fileName)
```
Comment (tvsonar) [inline src/Integration/LocalServices/ProjectSystemHelper.cs]: The name doesn't match the behavior. This not only removes the file from the project, but removes the project too if it was empty and its a solution folder.

I tend to think that adding a dedicated method to remove empty solution folders would be cleaner.

File: src/Integration/Persistence/SolutionBindingSerializer.cs [hunk 6]
```diff
@@ -97,17 +98,33 @@ private void AddSolutionItemFile(string configFile)
             var projectSystemHelper = this.serviceProvider.GetService<IProjectSystemHelper>();
             projectSystemHelper.AssertLocalServiceIsNotNull();
 
-            Project solutionItemsProject = projectSystemHelper.GetSolutionItemsProject();
+            Project solutionItemsProject = projectSystemHelper.GetSolutionFolderProject(Constants.SonarQubeManagedFolderName, true);
             if (solutionItemsProject == null)
             {
-                Debug.Fail("Could not find the solution items project");
+                Debug.Fail("Could not find the solution items project"); // Should never happen
```
Comment (tvsonar) [inline src/Integration/Persistence/SolutionBindingSerializer.cs]: I'd remove the comment. `Debug.Fail` already means that it should never happen. :-)

**Pull request merged**